scripts/src/join_metadata.R

```diff
@@ -0,0 +1,78 @@
+#' validate_columns_exist
+#' 
+#' This function checks that a list of columns are present in a dataframe.
+#' Columns that were not found in the dataframe are printed out.
+#' 
+#' @param selected_columns A vector of strings each representing a column name
+#' @param df A dataframe to check against
+#' @return Boolean
+validate_columns_exist= function(selected_cols, df) {
+  # Check that all of selected_columns are in df - base::setdiff(A, B) = A[!A %in% B].
+  unmatched_cols= base::setdiff(selected_cols, colnames(df))
+  
+  if(length(unmatched_cols) > 0) {
+    print('The following columns are missing: ')
+    print(unmatched_cols)
+    return(FALSE)
+  } else {
+    return(TRUE)
+  }
+}
+
+#' Join metadata
+#' 
+#' Joins a given data frame with the sample meta.
+#' 
+#' @import tidyverse
+#' @param input_df Input dataframe that should contain the columns specified in the "key_cols" parameter and "cell_set".
+#' @param metadata Dataframe of the sample meta used in the run.
+#' @param key_cols Vector of column names used as identifiers in the sample meta.
+#' @returns Data frame with additional columns from the sample meta.
+join_metadata= function(input_df, metadata, key_cols) {
+  require(tidyverse)
+  
+  # Validation: Check that key_cols are present in df ----
+  if(validate_columns_exist(key_cols, input_df) == FALSE) {
+    stop('Not all key_cols (printed above) are present in the provided dataframe.')
+  }
+  
+  # Validation: Check that key_cols are present in the sample meta ----
+  if(validate_columns_exist(key_cols, metadata) == FALSE) {
+    stop('Not all key_cols (printed above) are present in the provided metadata.')
+  }
+  
+  # Collapse the sample meta using key_cols and join onto the input df ----
+  # Collapse unique values into a single row and then filter out columns with the separator.
+  # Columns with only one unique value in a group are selected.
+  collapsed_metadata= metadata %>% dplyr::group_by(pick(all_of(key_cols))) %>% 
+    dplyr::summarise(across(everything(), function(x) paste(sort(unique(x)), collapse= ':::'))) %>% dplyr::ungroup() %>%
+    dplyr::select(all_of(key_cols), where(function(x) base::any(!grepl(':::', x))))
+  
+  # Join using the key_cols, drop any columns that were duplicated.
+  output_df= dplyr::left_join(input_df, collapsed_metadata, by= key_cols, 
+                              suffix= c('', '.y'), relationship='many-to-one') %>%
+    dplyr::select(-tidyselect::ends_with('.y'))
+  
+  # Validation: Check that merge did not explode ----
+  print(paste0('Input df rows: ', nrow(input_df)))
+  print(paste0('Output df rows: ', nrow(output_df)))
+  if(nrow(input_df) < nrow(output_df)) {
```

Comment:
Is it possible that  new rows are added because of the some data was removed using data to remove module?

Reply:
I don't think so. This is only in the join_metadata function to make sure that when joining the metadata, the function isn't creating more rows.